src/core/reference/include/ngraph/runtime/reference/grid_sample.hpp

```diff
@@ -0,0 +1,300 @@
+DATA_ET nearest(const DATA_ET* data,
+                const Shape& data_shape,
+                const size_t n,
+                const size_t c,
+                const GRID_ET y_n,
+                const GRID_ET x_n,
+                const get_padded_fn_t<DATA_ET>& get_padded,
+                const denormalize_fn_t<GRID_ET>& denormalize) {
+    const auto y_nearest = std::lrint(denormalize(y_n, data_shape[2]));
+    const auto x_nearest = std::lrint(denormalize(x_n, data_shape[3]));
+    return get_padded(data, data_shape, n, c, y_nearest, x_nearest);
+}
+
+template <typename T>
+using vector_4_t = typename std::array<T, 4>;
+template <typename T>
+using matrix_4x4_t = typename std::array<vector_4_t<T>, 4>;
+
+// formula taken from
+// https://github.com/onnx/onnx/blob/a92870cdf359297495a118184dca2eaecee4b717/onnx/backend/test/case/node/resize.py#L201-L207
+template <typename T>
+vector_4_t<T> cubic_coeffs(const T r, const T A = -0.75) {
+    vector_4_t<T> v;
+    v[0] = ((A * (r + 1) - 5 * A) * (r + 1) + 8 * A) * (r + 1) - 4 * A;
+    v[1] = ((A + 2) * r - (A + 3)) * r * r + 1;
+    v[2] = ((A + 2) * (1 - r) - (A + 3)) * (1 - r) * (1 - r) + 1;
+    v[3] = ((A * (2 - r) - 5 * A) * (2 - r) + 8 * A) * (2 - r) - 4 * A;
+    return v;
+}
+
+template <typename T>
+T scalar_prod(const vector_4_t<T>& a, const vector_4_t<T>& b) {
+    T c = 0;
+    for (int i = 0; i < a.size(); ++i)
+        c += a[i] * b[i];
+    return c;
+}
+
+template <typename DATA_ET>
+matrix_4x4_t<DATA_ET> gather_4x4(const DATA_ET* data,
+                                 const Shape& data_shape,
+                                 const size_t n,
+                                 const size_t c,
+                                 const long y_topleft,
+                                 const long x_topleft,
+                                 const get_padded_fn_t<DATA_ET>& get_padded) {
+    matrix_4x4_t<DATA_ET> s;
+    for (int j = 0; j < 4; ++j)
+        for (int i = 0; i < 4; ++i)
+            s[j][i] = get_padded(data, data_shape, n, c, y_topleft + j, x_topleft + i);
+    return s;
+}
+
+template <typename DATA_ET, typename GRID_ET>
+DATA_ET bicubic(const DATA_ET* data,
+                const Shape& data_shape,
+                const size_t n,
+                const size_t c,
+                const GRID_ET y_n,
+                const GRID_ET x_n,
+                const get_padded_fn_t<DATA_ET>& get_padded,
+                const denormalize_fn_t<GRID_ET>& denormalize) {
+    const auto y_d = denormalize(y_n, data_shape[2]);
+    const auto x_d = denormalize(x_n, data_shape[3]);
+    const auto y_topleft = std::floor(y_d);
+    const auto x_topleft = std::floor(x_d);
+    const auto dy = y_d - y_topleft;
+    const auto dx = x_d - x_topleft;
+    const auto s = gather_4x4(data, data_shape, n, c, y_topleft - 1, x_topleft - 1, get_padded);
+
+    const auto cy = cubic_coeffs(dy);
+    const auto cx = cubic_coeffs(dx);
+    vector_4_t<DATA_ET> p;
+    for (int i = 0; i < 4; ++i)
```

Comment:
This loop is equal to `std::transform` with `std::inner_product`.

Reply:
Done.